internal/gitea/repo.go

```diff
@@ -0,0 +1,216 @@
+// populates the information in r.Details.
+func (repo *Repository) processCommitsSinceRelease() error {
+	// TODO: This does not currently handle monorepos - see note about releases
+	// in `processReleases`. If there are releases in Gitea with a monorepo,
+	// then this function maybe needs to restrict the commits to ones where the
+	// tree overlaps - but maybe we actually want to know how many commits
+	// overall, because we want to count things like common code being adjusted.
+
+	if len(repo.Details.Releases) == 0 {
+		return errors.New("processCommitsSinceRelease must not be called without releases!")
+	}
+
+	// Add the commit delta between last release and default branch.
+	latestRelease := repo.Details.Releases[len(repo.Details.Releases)-1]
+	opts := gitea.ListCommitOptions{
+		ListOptions: gitea.ListOptions{PageSize: releasesPerRepo},
+		SHA:         latestRelease.Version,
+	}
+	source := repo.Details.Name
+	if repo.Details.Monorepo != "" {
+		source = repo.Details.Monorepo
+	}
+	commits, _, err := repo.client.ListRepoCommits(repo.org, source, opts)
+	if err != nil {
+		return err
+	}
+
+	repo.Details.NewCommits = len(commits)
+
+	return nil
+}
+
+// processCommits fetches the latest commits to a repository and populates them
+// into the repo struct in the case that there are no releases identified.
+func (repo *Repository) processCommits() error {
+	// TODO: This does not currently handle monorepos. It's not clear which
+	// commits should be counted in this case - only ones that have a tree that
+	// overlaps with the subfolder? All commits, as now? If there's common code,
+	// then all commits is probably truest to the single-repo meaning, but if
+	// there are commits that are only in a separate charm, then those probably
+	// should not be included.
+
+	opts := gitea.ListCommitOptions{
+		ListOptions: gitea.ListOptions{PageSize: releasesPerRepo},
+		SHA:         repo.defaultBranch,
+	}
+
+	source := repo.Details.Name
+	if repo.Details.Monorepo != "" {
+		source = repo.Details.Monorepo
+	}
+	commits, _, err := repo.client.ListRepoCommits(repo.org, source, opts)
+	if err != nil {
+		return err
+	}
+
+	// Iterate over the commits and append them to r.Details.Commits
+	for _, commit := range commits {
+		// Some commits don't have an author.
+		name := ""
+		if commit.Author != nil {
+			name = commit.Author.FullName
+		}
+		repo.Details.Commits = append(repo.Details.Commits, &repos.Commit{
+			Sha:       commit.CommitMeta.SHA,
+			Author:    name,
+			Timestamp: commit.CommitMeta.Created.Unix(),
+			Message:   renderReleaseBody(commit.RepoCommit.Message, repo),
+			URL:       commit.HTMLURL,
+		})
+	}
+
+	return nil
+}
+
+// renderReleaseBody transforms a Markdown string from a Gitea Release into HTML.
+func renderReleaseBody(body string, repo *Repository) string {
+	// Render the Markdown to HTML.
+	normalised := markdown.NormalizeNewlines([]byte(body))
+
+	return string(markdown.ToHTML(normalised, nil, nil))
+}
```

Comment:
This is duplicated somewhat from another package right?

It does include a bunch of issue/pr/user parsing for Github, but I wonder if we can avoid duplicating the code?

Reply:
There's a function in the `github` package with the same name and also has these two lines, yes. It's the one that also converts the GitHub shorthand into standalone HTML.

I haven't found any releases on opendev.org and I haven't been able to find any documentation about what format the release note is generally in, so markdown is an assumption here. It's probably best to just drop this entirely and stick with whatever type of string the `Release` response has, until there's evidence otherwise.